Add a `HeaderMap` to `HttpError` (#1193)
oxidecomputer/dropshot: CHANGELOG.adoc

returns a `ClientErrorStatusCode` if the status code is a client error, or an
error.

* `HttpError` now contains an `Option<Box<http::HeaderMap>>` of headers to add
to error responses constructed for the error.
+

> Code which constructs `HttpError` literals must now initialize this field. For
> example, this will no longer compile:
> +
> ```rust
> let err = dropshot::HttpError {
>     status_code: ErrorStatusCode::INTERNAL_SERVER_ERROR,
>     error_code: None
>     internal_message: "something bad happened".to_string(),
>     external_message: "Internal Error".to_string(),
> };
> ```
> +
> Instead, write:
> +
> ```rust
> let err = dropshot::HttpError {
>     status_code: ErrorStatusCode::INTERNAL_SERVER_ERROR,
>     error_code: None
>     internal_message: "something bad happened".to_string(),
>     external_message: "Internal Error".to_string(),
>     headers: None, // <-- Add this
> };
> ```

Reviewer: ```suggestion
Code that constructs `HttpError` literals must now initialize this field.
```

Author: Hm, @davepacheco had previously expressed a preference for including examples of breaking changes whenever possible...though I agree that this one is pretty trivial. Happy to remove it.

Author: Removed it in a34b66c1a2c2293a23cfa4ac3f3bd54effffe053.

+
Similarly, code which destructures `HttpError`s, such as in a `match` or `let`
expression, must now either destructure the `headers` field or add `..` to

> ignore it.

Reviewer: del

Author: IMO, this is worth mentioning, because otherwise, the language in the previous sentence suggests that this change _only_ applies to code that constructs a `HttpError` literal. We could simplify it to something like:

> Code which constructs a `HttpError` literal or destructures `HttpError`s in a pattern must be changed to add this field.

What do you think?

Reviewer: All this feels like the expected consequences of adding a new field to a structure. I don't expect folks will proactively fix things based on the changelog and I also don't think folks will go looking at the changelog to figure out how to address these compilation errors.

* `HttpError` now contains an `Option<Box<http::HeaderMap>>` of headers to add
to error responses constructed for the error.
+

> Code which constructs `HttpError` literals must now initialize this field.

Reviewer: ```suggestion
Code that constructs `HttpError` literals must now initialize this field.
```
